src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -422,473 +388,307 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
+
+            if (this.outputContext.WritingResponse)
+            {
+                this.ThrowODataException(Strings.ODataBatchWriter_CannotCreateRequestOperationWhenWritingResponse);
+            }
+
+            ExceptionUtils.CheckArgumentStringNotNullOrEmpty(method, "method");
+
+            ExceptionUtils.CheckArgumentNotNull(uri, "uri");
         }
 
         /// <summary>
-        /// Starts a new batch - implementation of the actual functionality.
+        /// Verifies that a call is allowed to the writer.
         /// </summary>
-        private void WriteStartBatchImplementation()
+        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
+        internal void VerifyCallAllowed(bool synchronousCall)
         {
-            this.SetState(BatchWriterState.BatchStarted);
+            if (synchronousCall)
+            {
+                if (!this.outputContext.Synchronous)
+                {
+                    throw new ODataException(Strings.ODataBatchWriter_SyncCallOnAsyncWriter);
+                }
+            }
+            else
+            {
+#if PORTABLELIB
+                if (this.outputContext.Synchronous)
+                {
+                    throw new ODataException(Strings.ODataBatchWriter_AsyncCallOnSyncWriter);
+                }
+#else
+                Debug.Assert(false, "Async calls are not allowed in this build.");
+#endif
+            }
         }
 
         /// <summary>
-        /// Verifies that calling WriteEndBatch is valid.
+        /// Catch any exception thrown by the action passed in; in the exception case move the writer into
+        /// state ExceptionThrown and then rethrow the exception.
         /// </summary>
-        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
-        private void VerifyCanWriteEndBatch(bool synchronousCall)
+        /// <param name="action">The action to execute.</param>
+        internal void InterceptException(Action action)
         {
-            this.ValidateWriterReady();
-            this.VerifyCallAllowed(synchronousCall);
+            try
+            {
+                action();
+            }
+            catch
+            {
+                if (!IsErrorState(this.state))
+                {
+                    this.SetState(BatchWriterState.Error);
+                }
+
+                throw;
+            }
         }
 
         /// <summary>
-        /// Ends a batch - implementation of the actual functionality.
+        /// Wrapper method to validate state transition with optional customized validation.
         /// </summary>
-        private void WriteEndBatchImplementation()
+        /// <param name="newState">Teh new writer state to transition into.</param>
+        /// <param name="customizedValidationAction">Optional validation action.</param>
+        internal void ValidateTransition(BatchWriterState newState, Action customizedValidationAction)

```

Comment:
ValidateTransition should be private and only called from the base class as part of state transition (facilitated by moving SetState to base class).